### PR TITLE
Add yolox fast loss to sg

### PR DESCRIPTION
**Pull request by shanibenbaruch**

### CODE REVIEW

File: src/super_gradients/training/losses/yolox_loss.py [hunk 1]
```diff
@@ -466,3 +468,362 @@ def dynamic_k_matching(self, cost, pair_wise_ious, gt_classes, num_gt, fg_mask):
...
+                    cls_preds[candidate_gt_img_ids, candidate_anchor_ids]
+                    .float().sigmoid_()
+                    * obj_preds[candidate_gt_img_ids, candidate_anchor_ids]
+                    .float().sigmoid_()
+            )
+            pair_wise_cls_cost = F.binary_cross_entropy(
+                cls_preds_.sqrt_(), gt_cls_per_image, reduction="none"
+            ).sum(-1)
+        ious_cost = -torch.log(candidate_ious + 1e-8)
+        cost = pair_wise_cls_cost + 3.0 * ious_cost + 100000.0 * strong_candidate_mask.logical_not()
```
Comment (ofrimasad) [inline src/super_gradients/training/losses/yolox_loss.py:765]: please make 3.0 and 100000.0 parameters of this function and of the function calling this and set them as defaults values in the compute_loss func

**Review by ofrimasad — COMMENTED** [review]
my comment inline

File: src/super_gradients/training/losses/yolox_loss.py [hunk 2]
```diff
@@ -466,3 +468,362 @@ def dynamic_k_matching(self, cost, pair_wise_ious, gt_classes, num_gt, fg_mask):
...
+    * Completely avoids using loops compared to the nested loops in the original implementation.
+        As a result runs much faster (speedup depends on the type of GPUs, their count, the batch size, etc.).
+    * Tensors format is very different the original implementation.
+        Tensors contain image ids, ground truth ids and anchor ids as values to support variable length data.
+    * There are differences in terms of the algorithm itself:
+    1. When computing a dynamic k for a ground truth,
+        in the original implementation they consider the sum of top 10 predictions sorted by ious among the initial
+        foregrounds of any ground truth in the image,
+        while in our implementation we consider only the initial foreground of that particular ground truth.
+        To compensate for that difference we introduce the dynamic_ks_bias hyperparamter which makes the dynamic ks larger.
```
[the remark this answers is not in the document]
Comment (shanibenbaruch) [reply src/super_gradients/training/losses/yolox_loss.py:487]: Yolox implementation uses other foregrounds of **any** gt in the image to calculate the dynamic k (they sum the best ious of 10 predictions). 
Ido's implementation sums the iou only on the foreground of the specific gt - most times it gives a smaller result of dynamic k.
to compensate for this he multiples by `dynamic_ks_bias=1.1`.

File: src/super_gradients/training/losses/yolox_loss.py [hunk 3]
```diff
@@ -466,3 +468,362 @@ def dynamic_k_matching(self, cost, pair_wise_ious, gt_classes, num_gt, fg_mask):
...
+    2. When computing the k matched detections for a ground truth,
+        in the original implementation they consider the initial foregrounds of any ground truth in the image as candidates,
+        while in our implementation we consider only the initial foreground of that particular ground truth as candidates.
+        We believe that this difference is minute.
+    """
+
+    def __init__(self, strides, im_size, num_classes, use_l1=False, center_sampling_radius=2.5, iou_type='iou',
+                 dynamic_ks_bias=1.1, sync_num_fgs=False, obj_loss_fix=False):
+        super().__init__(strides, im_size, num_classes, use_l1=use_l1, center_sampling_radius=center_sampling_radius,
```
[the remark this answers is not in the document]
Comment (shanibenbaruch) [reply src/super_gradients/training/losses/yolox_loss.py:506]: Added in d8a15699e725ea7ee6a776cac2e2473f81a7265f

File: src/super_gradients/training/losses/yolox_loss.py [hunk 4]
```diff
@@ -466,3 +468,362 @@ def dynamic_k_matching(self, cost, pair_wise_ious, gt_classes, num_gt, fg_mask):
...
+            candidate_anchor_ids), dim=1).unique(dim=0, return_inverse=True)[1]
+        return candidate_img_and_fg_ids_combined
+
+    def compute_dynamic_ks(self, ids: torch.Tensor, ious: torch.Tensor, dynamic_ks_bias) -> torch.Tensor:
+        """
+        :param ids:                 ids of GTs, shape: [num_candidates]
+        :param ious:                pairwise IoUs, shape: [num_candidates]
+        :param dynamic_ks_bias:     multiply the resulted k to compensate the regular loss
+        """
+        assert len(ids.shape) == 1
```
[the remark this answers is not in the document]
Comment (shanibenbaruch) [reply src/super_gradients/training/losses/yolox_loss.py]: Added in d8a15699e725ea7ee6a776cac2e2473f81a7265f

File: src/super_gradients/training/losses/yolox_loss.py (hunk 1 above)
Comment (shanibenbaruch) [reply src/super_gradients/training/losses/yolox_loss.py]: I added them in the __init__ of YoloXDetectionLoss, so it can be used for the regular loss as well.
https://github.com/Deci-AI/super-gradients/blob/d8a15699e725ea7ee6a776cac2e2473f81a7265f/src/super_gradients/training/losses/yolox_loss.py#L74
